### PR TITLE
EIP1:5626: Summary view repository and mappers + Liquibase 

### PR DESCRIPTION
**Pull request by vishalgupta-valtech**
- New view updated to include `id` of `anonymous_elector_document` entity
- Added repository and it's int test
- Added mapper that will be used by service and controller in subsequent PRs

### CODE REVIEW

File: src/main/kotlin/uk/gov/dluhc/printapi/database/repository/AnonymousElectorDocumentSummaryRepository.kt [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+import org.springframework.data.repository.PagingAndSortingRepository
+import org.springframework.stereotype.Repository
+import uk.gov.dluhc.printapi.database.entity.AnonymousElectorDocumentSummary
+import uk.gov.dluhc.printapi.database.entity.SourceType
+import java.util.UUID
+
+@Repository
+interface AnonymousElectorDocumentSummaryRepository :
+    PagingAndSortingRepository<AnonymousElectorDocumentSummary, UUID> {
```
Comment (vishalgupta-valtech) [inline src/main/kotlin/uk/gov/dluhc/printapi/database/repository/AnonymousElectorDocumentSummaryRepository.kt:11]: I had to add `id` column in the view (See liquibase scripts)

File: src/main/resources/db/changelog/ddl/0026_update_aed_summary_view.xml [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+            addr.postcode
+            FROM anonymous_elector_document aed
+            JOIN aed_contact_details contact ON aed.aed_contact_details_id = contact.id
+            JOIN address addr on contact.address_id = addr.id
+            JOIN (
+                    SELECT
+                        gss_code,
+                        source_type,
+                        source_reference,
+                        max(request_date_time) as mostRecentRequestDateTime
```
Comment (matt-wills-valtech) [inline src/main/resources/db/changelog/ddl/0026_update_aed_summary_view.xml:34]: Very minor - why the change in style from underscores to camel case?

File: src/test/kotlin/uk/gov/dluhc/printapi/mapper/aed/AnonymousSearchSummaryMapperTest.kt [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+@ExtendWith(MockitoExtension::class)
+class AnonymousSearchSummaryMapperTest {
+
+    @Mock
+    private lateinit var instantMapper: InstantMapper
+
+    @InjectMocks
+    private lateinit var mapper: AnonymousSearchSummaryMapperImpl
+
+    @Nested
```
Comment (NathanRussellValtech) [inline src/test/kotlin/uk/gov/dluhc/printapi/mapper/aed/AnonymousSearchSummaryMapperTest.kt:30]: Minor ... question over inner classes like this ....
I understand the value inner classes has when there are multiple test methods grouped in the inner class, but in this case we dont ... 🤔 

FWIW I've just ripped out an inner class in the mapper test I am currently working on for this very reason ....

File: src/main/kotlin/uk/gov/dluhc/printapi/database/entity/AnonymousElectorDocumentSummary.kt [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+import javax.persistence.Id
+import javax.persistence.Table
+
+@Table(name = "v_anonymous_elector_document_summary")
+@Entity
+class AnonymousElectorDocumentSummary(
+
+    @Id
+    @Type(type = UUIDCharType)
+    val id: UUID,
```
Comment (matt-wills-valtech) [inline src/main/kotlin/uk/gov/dluhc/printapi/database/entity/AnonymousElectorDocumentSummary.kt:21]: This is the ID of the AED right?

**Review by NathanRussellValtech — APPROVED** [review]
Minor comment on tests & inner classes, but otherwise LGTM 👍

File: src/test/kotlin/uk/gov/dluhc/printapi/database/repository/AnonymousElectorDocumentSummaryRepositoryIntegrationTest.kt [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+        @Test
+        fun `should find all AEDs for a given gssCode and source type pages`() {
+            // Given
+            val gssCode = aGssCode()
+            val currentDate = LocalDate.now()
+            val currentDateTimeInstant = Instant.now()
+            val aed1SourceReference = aValidSourceReference()
+            val aed1ApplicationReference = aValidApplicationReference()
+
+            val aed1FirstCertificate = buildAnonymousElectorDocument(
```
Comment (matt-wills-valtech) [inline src/test/kotlin/uk/gov/dluhc/printapi/database/repository/AnonymousElectorDocumentSummaryRepositoryIntegrationTest.kt:70]: Certificate in the names? I'm not sure I understand the names of these variables. Also, this is a large given section, which makes it difficult to grasp the intention of the test.

File: src/test/kotlin/uk/gov/dluhc/printapi/database/repository/AnonymousElectorDocumentSummaryRepositoryIntegrationTest.kt [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+                    sourceType = ANONYMOUS_ELECTOR_DOCUMENT,
+                    pageRequest = withPageRequestAndSortOrder(0, 100)
+                )
+
+            // Then
+            assertThat(actual).isNotNull.isEmpty()
+        }
+
+        @Test
+        fun `should find all AEDs for a given gssCode and source type pages`() {
```
Comment (matt-wills-valtech) [inline src/test/kotlin/uk/gov/dluhc/printapi/database/repository/AnonymousElectorDocumentSummaryRepositoryIntegrationTest.kt:62]: Why 'pages' in the name?

File: src/test/kotlin/uk/gov/dluhc/printapi/database/repository/AnonymousElectorDocumentSummaryRepositoryIntegrationTest.kt [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+            val expectedSummaryRecord3 = buildAnonymousElectorDocumentSummaryViewFromAedEntity(aed1LatestCertificate)
+            val expectedSummaryRecord4 = buildAnonymousElectorDocumentSummaryViewFromAedEntity(aed2LatestCertificate)
+            val otherAedSummaryRecord = buildAnonymousElectorDocumentSummaryViewFromAedEntity(otherAed)
+
+            // When
+            val actual = anonymousElectorDocumentSummaryRepository
+                .findAllByGssCodeInAndSourceType(
+                    gssCodes = listOf(gssCode),
+                    sourceType = ANONYMOUS_ELECTOR_DOCUMENT,
+                    pageRequest = withPageRequestAndSortOrder(0, 100)
```
Comment (matt-wills-valtech) [inline src/test/kotlin/uk/gov/dluhc/printapi/database/repository/AnonymousElectorDocumentSummaryRepositoryIntegrationTest.kt:144]: Ok, so we're testing some pagination behaviour, despite the fact it's not supported via the API yet?

I wonder if this is trying to test too many concerns and could be split up a bit?

Comment (vishalgupta-valtech) [reply src/test/kotlin/uk/gov/dluhc/printapi/database/repository/AnonymousElectorDocumentSummaryRepositoryIntegrationTest.kt:144]: The repository requires pageRequest, so added this test.

File: src/main/resources/db/changelog/ddl/0026_update_aed_summary_view.xml [hunk 8]
```diff
@@ -0,0 +1,81 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<databaseChangeLog
+    xmlns="http://www.liquibase.org/xml/ns/dbchangelog"
+    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog
+      http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-4.1.xsd">
+
+    <changeSet author="vishal.gupta@valtech.com" id="0026_update_aed_summary_view - Adds id to AED summary view" context="ddl">
```
Comment (matt-wills-valtech) [inline src/main/resources/db/changelog/ddl/0026_update_aed_summary_view.xml:8]: You know you want to make this 0027! https://github.com/cabinetoffice/eip-ero-print-api/pull/237 😆 😉

Comment (vishalgupta-valtech) [reply src/main/resources/db/changelog/ddl/0026_update_aed_summary_view.xml:8]: Technically this PR was raised before #237  😜 but If you want I can wait for your PR to be merged first

Comment (matt-wills-valtech) [reply src/main/resources/db/changelog/ddl/0026_update_aed_summary_view.xml:8]: I don't mind. It depends on which one gets merged first 😄

File: src/test/kotlin/uk/gov/dluhc/printapi/database/repository/AnonymousElectorDocumentSummaryRepositoryIntegrationTest.kt [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+                )
+
+            // Then
+            assertThat(actual).isNotNull.isEmpty()
+        }
+
+        @Test
+        fun `should returns empty list when no matching gssCode record exists`() {
```
Comment (matt-wills-valtech) [inline src/test/kotlin/uk/gov/dluhc/printapi/database/repository/AnonymousElectorDocumentSummaryRepositoryIntegrationTest.kt:43]: Minor - typo on `returns`

File: src/test/kotlin/uk/gov/dluhc/printapi/database/repository/AnonymousElectorDocumentSummaryRepositoryIntegrationTest.kt [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+            )
+            val aed2LatestDocument = buildAnonymousElectorDocument(
+                gssCode = gssCode,
+                sourceReference = aed2SourceReference,
+                applicationReference = aed2ApplicationReference,
+                issueDate = currentDate.minusDays(9),
+                requestDateTime = currentDateTimeInstant, // View will return this latest record
+                contactDetails = aed2InitialDocument.contactDetails!!
+            )
+            val aed3Document = buildAnonymousElectorDocument(gssCode = gssCode, issueDate = currentDate.plusDays(10))
```
Comment (matt-wills-valtech) [inline src/test/kotlin/uk/gov/dluhc/printapi/database/repository/AnonymousElectorDocumentSummaryRepositoryIntegrationTest.kt:121]: plusDays(10) seems odd since it will never happen in practice?

File: src/test/kotlin/uk/gov/dluhc/printapi/database/repository/AnonymousElectorDocumentSummaryRepositoryIntegrationTest.kt [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+            val actual = anonymousElectorDocumentSummaryRepository
+                .findAllByGssCodeInAndSourceType(
+                    gssCodes = listOf(gssCode),
+                    sourceType = ANONYMOUS_ELECTOR_DOCUMENT,
+                    pageRequest = withPageRequestAndSortOrder(0, 100)
+                )
+
+            // Then
+            assertThat(actual)
+                .doesNotContain(otherAedSummaryRecord)
```
Comment (matt-wills-valtech) [inline src/test/kotlin/uk/gov/dluhc/printapi/database/repository/AnonymousElectorDocumentSummaryRepositoryIntegrationTest.kt:149]: Nice 👍

**Review by matt-wills-valtech — APPROVED** [review]
LGTM. Just a couple of minor comments.

File: src/main/resources/db/changelog/ddl/0026_update_aed_summary_view.xml (hunk 8 above)
Comment (matt-wills-valtech) [reply src/main/resources/db/changelog/ddl/0026_update_aed_summary_view.xml:8]: I've renamed mine to 27 👍

**Review by matt-wills-valtech — APPROVED** [review]

File: src/test/kotlin/uk/gov/dluhc/printapi/database/repository/AnonymousElectorDocumentSummaryRepositoryIntegrationTest.kt (hunk 10 above)
Comment (vishalgupta-valtech) [reply src/test/kotlin/uk/gov/dluhc/printapi/database/repository/AnonymousElectorDocumentSummaryRepositoryIntegrationTest.kt]: I have renamed the variables now. This is for another application that was added after 10 days, so is very much possible. I think the confusion could have been because the variable name was earlier renamed with "world class" standards 😆
